File: app/src/main/java/io/oldering/tvfoot/red/app/domain/matches/state/MatchesViewState.java
package io.oldering.tvfoot.red.app.domain.matches.state;

import com.google.auto.value.AutoValue;
import io.oldering.tvfoot.red.app.domain.matches.displayable.HeaderRowDisplayable;
import io.oldering.tvfoot.red.app.domain.matches.displayable.LoadingRowDisplayable;
import io.oldering.tvfoot.red.app.domain.matches.displayable.MatchRowDisplayable;
import io.oldering.tvfoot.red.app.domain.matches.displayable.MatchesItemDisplayable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import javax.annotation.Nullable;

@AutoValue public abstract class MatchesViewState {
  public List<MatchesItemDisplayable> matchesItemDisplayables() {
    List<String> headers = new ArrayList<>();
    List<MatchesItemDisplayable> items = new ArrayList<>();
    for (MatchRowDisplayable match : matches()) {
      if (!headers.contains(match.headerKey())) {
        headers.add(match.headerKey());
        items.add(HeaderRowDisplayable.create(match.headerKey()));
      }
      items.add(match);
    }
    if (!items.isEmpty()) {
      items.add(LoadingRowDisplayable.create());
    }
    return items;
  }

  public abstract List<MatchRowDisplayable> matches();

  @Nullable public abstract Throwable error();

  public abstract boolean nextPageLoading();

  public abstract boolean refreshLoading();

  @Nullable public abstract MatchRowDisplayable clickedMatch();

  public abstract Integer currentPage();

  public static Builder builder() {
    return new AutoValue_MatchesViewState.Builder().matches(Collections.emptyList())
        .error(null)
        .nextPageLoading(false)
        .refreshLoading(false)
        .clickedMatch(null)
        .currentPage(0);
  }

  public abstract Builder buildWith();

  static MatchesViewState idle() {
    return MatchesViewState.builder().build();
  }

  @AutoValue.Builder public static abstract class Builder {
    public abstract Builder matches(List<MatchRowDisplayable> matches);

    public abstract Builder error(@Nullable Throwable error);

    public abstract Builder nextPageLoading(boolean nextPageLoading);

    public abstract Builder refreshLoading(boolean pullToRefreshLoading);

    public abstract Builder clickedMatch(@Nullable MatchRowDisplayable clickedMatch);

    public abstract Builder currentPage(Integer CurrentPage);

    public abstract MatchesViewState build();
  }
}

File: app/src/main/java/com/benoitquenaudon/tvfoot/red/app/domain/matches/MatchesViewState.kt
package com.benoitquenaudon.tvfoot.red.app.domain.matches

import com.benoitquenaudon.tvfoot.red.app.data.entity.FilterTeam
import com.benoitquenaudon.tvfoot.red.app.data.entity.Tag
import com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable.HeaderRowDisplayable
import com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable.LoadingRowDisplayable
import com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable.MatchRowDisplayable
import com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable.MatchesItemDisplayable
import com.benoitquenaudon.tvfoot.red.app.domain.matches.filters.FiltersItemDisplayable.TeamSearchResultDisplayable
import com.benoitquenaudon.tvfoot.red.app.mvi.MviViewState
import com.benoitquenaudon.tvfoot.red.util.TagName
import com.benoitquenaudon.tvfoot.red.util.TagTargets
import com.benoitquenaudon.tvfoot.red.util.TeamCode
import java.util.ArrayList

data class MatchesViewState(
    val matches: List<MatchRowDisplayable> = emptyList(),
    val error: Throwable? = null,
    val nextPageLoading: Boolean = false,
    val refreshLoading: Boolean = false,
    val currentPage: Int = 0,
    val hasMore: Boolean = true,
    val tagsLoading: Boolean = false,
    val tagsError: Throwable? = null,
    val tags: List<Tag> = emptyList(),
    var filteredTags: Map<TagName, TagTargets> = emptyMap(),
    val searchInput: String = "",
    val searchingTeam: Boolean = false,
    val searchedTeams: List<TeamSearchResultDisplayable> = emptyList(),
    val teams: List<FilterTeam> = emptyList(),
    val filteredTeams: List<TeamCode> = emptyList()
) : MviViewState {
  fun matchesItemDisplayables(
      hasMore: Boolean,
      filteredTags: Map<String, List<String>>,
      filteredTeams: List<TeamCode>
  ): List<MatchesItemDisplayable> {

    val headers = ArrayList<String>()
    val items = ArrayList<MatchesItemDisplayable>()

    val filteredTargets = filteredTags.values.flatten().toSet()
    val filteredTeamsCodes = filteredTeams.toSet()
    for (match in matches) {
      if (filteredTargets.isNotEmpty() && filteredTargets.intersect(match.tags).isEmpty()) {
        continue
      }
      if (filteredTeamsCodes.isNotEmpty() &&
          filteredTeamsCodes.intersect(
              setOf(match.homeTeam.code, match.awayTeam.code)).isEmpty()) {
        continue
      }

      if (!headers.contains(match.headerKey)) {
        headers.add(match.headerKey)
        items.add(HeaderRowDisplayable.create(match.headerKey))
      }
      items.add(match)
    }
    if (!items.isEmpty() && hasMore) {
      items.add(LoadingRowDisplayable)
    }
    return items
  }

  companion object {
    fun idle(): MatchesViewState = MatchesViewState()
  }
}
